common/djangoapps/student/views.py

```diff
@@ -520,6 +515,13 @@ def dashboard(request):
         course_enrollment_pairs, course_modes_by_course
     )
 
+    # Retrieve the course modes for each course
+    enrolled_courses_dict = {}
+    for course, __ in course_enrollment_pairs:
+        enrolled_courses_dict[unicode(course.id)] = course
+
+    credit_messages = _create_credit_availability_message(enrolled_courses_dict, user)
```

Comment:
If you haven't already, could you check how many additional database queries this is executing on the dashboard?  Ideally the number of queries should be constant with respect to the number of enrollments.